### PR TITLE
refactor and add format plugin system

### LINKED ISSUE #164: [Feature Request] Format plugin system
**Summary**

It's not a good idea to put all specific format functions into the main class. We should implement a plugin system like ASE's calculator or so on, where the main class can register the plugin class.

**Detailed Description**

**Further Information, Files, and Links**

### PR DESCRIPTION
**Pull request by njzjz**
fix #164.

### CODE REVIEW

File: dpdata/plugins/gromacs.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import dpdata.gromacs.gro
+from dpdata.format import Format
+
+
+@Format.register("gro")
+@Format.register("gromacs/gro")
+class PwmatOutputFormat(Format):
```
Comment (amcadmus) [inline dpdata/plugins/gromacs.py:7]: Incorrect class name. Shouldn't be something like `GromacsGroFormat`?

File: dpdata/plugins/xyz.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from dpdata.xyz.quip_gap_xyz import QuipGapxyzSystems
+from dpdata.format import Format
+
+
+@Format.register("quip/gap/xyz")
+@Format.register("quip/gap/xyz_file")
+class DeePMDRawFormat(Format):
```
Comment (amcadmus) [inline dpdata/plugins/xyz.py:7]: Something like `QuipGapXYZ`?

File: plugin_example/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+> Data Summary
+> Unlabeled System
+> -------------------
+> Frame Numbers     : 12
+> Atom Numbers      : 20
+> Element List      :
+> -------------------
+> X
+> 20
```
Comment (amcadmus) [inline plugin_example/README.md:22]: Should be quoted with ```

File: plugin_example/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# An example for plugin
```
Comment (amcadmus) [inline plugin_example/README.md:1]: How about adding a reference in the main README to this example?

File: dpdata/plugins/rdkit.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+from dpdata.format import Format
+try:
+    import rdkit.Chem
+except ModuleNotFoundError:
+    pass
+import dpdata.rdkit.utils
+
+
+@Format.register("mol")
+@Format.register("mol_file")
```
Comment (Ericwang6) [inline dpdata/plugins/rdkit.py:10]: Why here are 2 register decorators?

Comment (njzjz) [reply dpdata/plugins/rdkit.py:10]: All of the following usage will be supported:
* `from_mol(file_name)`
* `from_mol_file(file_name)`
* `to_mol(file_name)`
* `to_mol_file(file_name)`
* `BondOrderSystem(file_name, fmt='mol')`
* `BondOrderSystem(file_name, fmt='mol_file')`
* `to('mol', file_name)`
* `to('mol_file', file_name)`

**Review by amcadmus — APPROVED** [review]

**Comment by njzjz** [comment]
It seems that I chose the wrong branch? We need to sync `master` with `devel`.

**Comment by amcadmus** [comment]
> It seems that I chose the wrong branch? We need to sync `master` with `devel`.

Yes you are right...

**Pull request merged**